autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java

```diff
@@ -0,0 +1,187 @@
+                    stackResponseUtils.getHostGroupInstanceType(stackV4Response, loadAlert.getScalingPolicy().getHostGroup());
+
+            YarnScalingServiceV1Response yarnResponse = yarnMetricsClient
+                    .getYarnMetricsForCluster(cluster, hostGroupInstanceType, stackV4Response.getCloudPlatform());
+
+            Map<String, String> hostFqdnsToInstanceId = stackResponseUtils
+                    .getCloudInstanceIdsForHostGroup(stackV4Response, loadAlert.getScalingPolicy().getHostGroup());
+            yarnResponse.getScaleUpCandidates().ifPresentOrElse(
+                    scaleUpCandidates -> handleScaleUp(hostGroupInstanceType, scaleUpCandidates, hostFqdnsToInstanceId.size()),
+                    () -> {
+                        yarnResponse.getScaleDownCandidates().ifPresent(
+                                scaleDownCandidates -> handleScaleDown(scaleDownCandidates, hostFqdnsToInstanceId));
+                    });
+
+            LOGGER.debug("Finished loadEvaluator for cluster {} in {} ms", cluster.getStackCrn(), System.currentTimeMillis() - start);
+        } catch (Exception ex) {
+            LOGGER.info("Failed to process load alert for Cluster {}, exception {}", cluster.getStackCrn(), ex);
+            eventPublisher.publishEvent(new UpdateFailedEvent(clusterId));
+        } finally {
+            LOGGER.debug("Finished loadEvaluator for cluster {} in {} ms", cluster.getStackCrn(), System.currentTimeMillis() - start);
+        }
+    }
+
+    protected void handleScaleUp(String hostGroupInstanceType, NewNodeManagerCandidates newNMCandidates, Integer existingHostGroupSize) {
+        Integer yarnRecommendedHostGroupCount =
+                newNMCandidates.getCandidates().stream()
+                        .filter(candidate -> candidate.getModelName().equalsIgnoreCase(hostGroupInstanceType))
+                        .findFirst()
+                        .map(NewNodeManagerCandidates.Candidate::getCount)
+                        .orElseThrow(() -> new RuntimeException(String.format(
+                                "Yarn Scaling API Response does not contain recommended node count " +
+                                        " for hostGroupInstanceType %s in Cluster %s, Yarn Response %s",
+                                        hostGroupInstanceType, cluster.getStackCrn(), newNMCandidates)));
+
+        Integer maxAllowedScaleUp = loadAlert.getLoadAlertConfiguration().getMaxResourceValue() - existingHostGroupSize;
+        Integer scaleUpCount = IntStream.of(yarnRecommendedHostGroupCount, DEFAULT_MAX_SCALE_UP_STEP_SIZE, maxAllowedScaleUp)
+                .filter(i -> i >= 0)
+                .min()
+                .getAsInt();
+
+        LOGGER.info("ScaleUp NodeCount {} for Cluster {}, HostGroup {]", scaleUpCount,
+                cluster.getStackCrn(), loadAlert.getScalingPolicy().getHostGroup());
+
+        if (scaleUpCount > 0) {
+            loadAlert.getScalingPolicy().setScalingAdjustment(scaleUpCount);
```

Comment:
The same comment about modifying a 'Model' object applies here. Is there any chance this gets persisted to the DB accidentally?
The 'scaleUpCount' is essentially the same as the 'decommissionNodeIds' during scaleDown. Why not move this into ScalingRequest as well?

For all the other Evaluators, ScalingAdjustment is static, and is not modified by the Evaluators.

Reply:
ok